### ISSUE TITLE
Client Freeze / Crash on attaching Angled Gear

### ISSUE THREAD

**Issue by Konfuz** [anegostudios/VintageStory-Issues#1448]
**Game Version:** v1.15.10  
**Platform:** Linux (manjaro)
**Modded:** Yes
**SP/MP:** Multiplayer

 <details><summary>Installed Modfiles</summary>

```
3.12.2021 11:54:39 [Notification] Started 62 systems on Client:
3.12.2021 11:54:39 [Notification]     Mod 'VSEssentials.dll' (game):
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.Core
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.EntityPartitioning
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ErrorReporter
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.ModCompatiblityUtil
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.NoObf.ModJsonPatchLoader
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.FallingBlockParticlesModSystem
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.CharacterExtraDialogs
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.EntityNameTagRendererRegistry
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModSystemHandbook
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.RoomRegistry
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.WeatherSystemCommands
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.WeatherSystemClient
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.WorldMapManager
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.LoadColorMaps
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ClothManager
3.12.2021 11:54:39 [Notification]     Mod 'extraoverlays' (extraoverlays):
3.12.2021 11:54:39 [Notification]         SharedUtils.ConstantsCore
3.12.2021 11:54:39 [Notification]         SharedUtils.Extensions.LoggerNetwork
3.12.2021 11:54:39 [Notification]         MoreOverlays.Core
3.12.2021 11:54:39 [Notification]     Mod 'VSSurvivalMod.dll' (survival):
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.SaplingControl
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.SurvivalCoreSystem
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.ModSystemRiftWeather
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.GenFromHeightmap
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.GenMaps
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.DebugSystem
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.UpgradeTasks
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.CharacterSystem
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.FruitingSystem
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModTemperature
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.SystemTemporalStability
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.MealMeshCache
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModSystemWearableStats
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModSystemBlockReinforcement
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ChiselBlockModelCache
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.GameEffects
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModJournal
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModLootRandomizer
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.ModSleeping
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.TeleporterManager
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.Mechanics.MechanicalPowerMod
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.GenStructures
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.GenStructuresPosPass
3.12.2021 11:54:39 [Notification]         Vintagestory.GameContent.TraderOutfits
3.12.2021 11:54:39 [Notification]     Mod 'VSCreativeMod.dll' (creative):
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.Core
3.12.2021 11:54:39 [Notification]         Vintagestory.ServerMods.WorldEdit.WorldEdit
3.12.2021 11:54:39 [Notification]     Mod 'VSHUD.dll' (vshud):
3.12.2021 11:54:39 [Notification]         VSHUD.ConfigLoader
3.12.2021 11:54:39 [Notification]         VSHUD.ModSystemNotification
3.12.2021 11:54:39 [Notification]         VSHUD.FloatyDamage
3.12.2021 11:54:39 [Notification]         VSHUD.ClaimUI
3.12.2021 11:54:39 [Notification]         VSHUD.MacroCommands
3.12.2021 11:54:39 [Notification]         VSHUD.MeshTools
3.12.2021 11:54:39 [Notification]         VSHUD.PlacementPreview
3.12.2021 11:54:39 [Notification]         VSHUD.ClockDialogSystem
3.12.2021 11:54:39 [Notification]         VSHUD.LightUtilModSystem
3.12.2021 11:54:39 [Notification]         VSHUD.TapeMeasure
3.12.2021 11:54:39 [Notification]         VSHUD.WaypointUtils
3.12.2021 11:54:39 [Notification]         VSHUD.VanillaPatches
3.12.2021 11:54:39 [Notification]         VSHUD.ClientModSystem
3.12.2021 11:54:39 [Notification]         VSHUD.VSHUDMain
3.12.2021 11:54:39 [Notification]         VSHUD.CommandRegistry
3.12.2021 11:54:39 [Notification]     Mod 'apacheTechWaypointExtensions' (wpex):
3.12.2021 11:54:39 [Notification]         ApacheTech.WaypointExtensions.Mod.ModSystems.WpexModSystem
3.12.2021 11:54:39 [Notification]     Mod 'StepUp-v1.2.0.cs' (stepup):
3.12.2021 11:54:39 [Notification]         StepUp.StepUpSystem
3.12.2021 11:54:39 [Notification]     Mod 'automap' (automap):
3.12.2021 11:54:39 [Notification]         Automap.AutomapMod
```
</details>

### Description
Freeze on attaching an Angled Gear to an axle adjacent to a large gear.

Its a rather ugly crash too, since VS keeps dips on the mouse-pointer and parts of the gnome-session are unavailiable while VS remains frozen. Terminating with SIGTERM doesnt work, I needed to use SIGKILL for your process to die - after which my system continued to work regulary

### How to reproduce
<strike>I can reproduce it with that specific gear in that specific place, but</strike> cannot provide you a snapshot since I don't own the server. Hopefully the logfile will be enough for you. 

### Logs
[client-crash.txt](https://github.com/anegostudios/VintageStory-Issues/files/7649009/client-crash.txt)

### Edit
After reconnecting once more I was unable to add another gear since the game thought there was already a node in the way. I had to deconstruct the large gear and after placing it again I was able to (re-)attach all required angled gears without issues.

**Comment by copygirl** [comment]
Pasting crash for reference:
```
Version: v1.15.10 (Stable)12/3/2021 11:56:30 AM: Critical error occurred
System.NullReferenceException: Object reference not set to an instance of an object
  at Vintagestory.GameContent.Mechanics.BEBehaviorMPBase.JoinAndSpreadNetworkToNeighbours (Vintagestory.API.Common.ICoreAPI api, Vintagestory.GameContent.Mechanics.MechanicalNetwork network, Vintagestory.GameContent.Mechanics.MechPowerPath exitTurnDir, Vintagestory.API.MathTools.Vec3i& missingChunkPos) [0x001a2] in <e6e85b96b1d6495cb901a26788ae32ec>:0 
  at Vintagestory.GameContent.Mechanics.BEBehaviorMPBase.spreadTo (Vintagestory.API.Common.ICoreAPI api, Vintagestory.GameContent.Mechanics.MechanicalNetwork network, Vintagestory.API.MathTools.BlockPos exitPos, Vintagestory.GameContent.Mechanics.MechPowerPath propagatePath, Vintagestory.API.MathTools.Vec3i& missingChunkPos) [0x0010c] in <e6e85b96b1d6495cb901a26788ae32ec>:0 
  at Vintagestory.GameContent.Mechanics.BEBehaviorMPBase.tryConnect (Vintagestory.API.MathTools.BlockFacing toFacing) [0x001d4] in <e6e85b96b1d6495cb901a26788ae32ec>:0 
  at Vintagestory.GameContent.Mechanics.BlockAngledGears.TryPlaceBlock (Vintagestory.API.Common.IWorldAccessor world, Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.ItemStack itemstack, Vintagestory.API.Common.BlockSelection blockSel, System.String& failureCode) [0x0021e] in <e6e85b96b1d6495cb901a26788ae32ec>:0 
  at Vintagestory.Client.NoObf.ClientMain.OnPlayerTryPlace (Vintagestory.API.Common.BlockSelection blockSelection, System.String& failureCode) [0x0005d] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.OnBlockBuild (Vintagestory.API.Common.BlockSelection blockSelection, Vintagestory.API.Common.Block onBlock, System.String& failureCode) [0x00059] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.HandleMouseInteractionsBlockSelected (System.Single dt) [0x0032b] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.UpdatePicking (System.Single dt) [0x0009e] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.OnFinalizeFrame (System.Single dt) [0x0004a] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.API.Client.DummyRenderer.OnRenderFrame (System.Single deltaTime, Vintagestory.API.Client.EnumRenderStage stage) [0x00000] in <1c7c8ad300544797b5d9d0b6d63dbb03>:0 
  at Vintagestory.Client.NoObf.ClientEventManager.TriggerRenderStage (Vintagestory.API.Client.EnumRenderStage stage, System.Single dt) [0x00021] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.ClientMain.TriggerRenderStage (Vintagestory.API.Client.EnumRenderStage stage, System.Single dt) [0x00028] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.ClientMain.RenderToDefaultFramebuffer (System.Single dt) [0x00092] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at _37NyQC3j7xjnAt2h09BqVJtTM8J._oYniXsjcVAWxLpTiLUivE84wTkA (System.Single ) [0x0003e] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at _R2SaBRGSTE9aIRQAjhFb2LrMpJb._gq2ZjDtIZ3sLq4OCUiRUzkjv3XA (System.Single ) [0x001b4] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at _R2SaBRGSTE9aIRQAjhFb2LrMpJb._tjdmYyEuuoCShcgREvOOHIl2K1G (System.Single ) [0x000cc] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame (System.Object sender, OpenTK.FrameEventArgs e) [0x00141] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
  at (wrapper delegate-invoke) System.EventHandler`1[OpenTK.FrameEventArgs].invoke_void_object_TEventArgs(object,OpenTK.FrameEventArgs)
  at OpenTK.GameWindow.OnRenderFrame (OpenTK.FrameEventArgs e) [0x00000] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at OpenTK.GameWindow.OnRenderFrameInternal (OpenTK.FrameEventArgs e) [0x00010] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at OpenTK.GameWindow.RaiseRenderFrame (System.Double elapsed, System.Double& timestamp) [0x0000c] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at OpenTK.GameWindow.DispatchRenderFrame () [0x00038] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at OpenTK.GameWindow.Run (System.Double updates_per_second, System.Double frames_per_second) [0x00104] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at OpenTK.GameWindow.Run () [0x00000] in <90b019caed434e81aee7f93ee8c94ce2>:0 
  at _rwEcecUjMyzMSfVPsyS8Wx2kGqk._XiZO0qHpODOJko3v6YCZBF7Z5Se (_2kddpZuU7wcyt2belKg18dkvHxo , System.String[] ) [0x003e2] in <5d85a0e633f241e8bf1526f1dee7cfe9>:0 
  at _rwEcecUjMyzMSfVPsyS8Wx2kGqk+_wCtdAYOq4Mvqin5W2bIeY37ND8u._hsN2K8eeMYDyaBUdgzJXpcVB22X () [0x00000] in <5d85a0e633f241e8bf1526f1dee7cfe9>:0 
  at _E5WqgB878qvAOVUQ9yfqu340l7G._XiZO0qHpODOJko3v6YCZBF7Z5Se (System.Threading.ThreadStart ) [0x00007] in <be10c4c7e7ee4ac5b68783f6507c58a7>:0 
```

**Comment by Craluminum2413** [comment]
Try to disable VSHUD mod

**Comment by Konfuz** [comment]
> Try to disable VSHUD mod

its not in the Stacktrace copygirl just copied :dancers:, how would VSHUD affect this?

**Comment by copygirl** [comment]
It is possible for a mod to modify methods from the base game using the library "Harmony", and currently there is no way to see that in a stacktrace. If Craluminum is talking from experience, I do suggest to try it without VSHUD.

**Comment by Konfuz** [comment]
> It is possible for a mod to modify methods from the base game using the library "Harmony", and currently there is no way to see that in a stacktrace. If Craluminum is talking from experience, I do suggest to try it without VSHUD.

Sounds a bit dirty & dangerous. But thanks for the explanation. Im not sure I got it accross well: I am no longer able to reproduce, it crashed two times in a row, then just refused to accept angled gears and now the large gear in question is fixed. 

Also it seems to have a reasonably low probability of occuring - I had large gears with wonky behavior before (not accepting gears in a specific place) but this hard freeze with system-impact was the first time I observed it. I'll keep you posted if I encounter the issue again (with VSHUD enabled) - is there any debug I should enable just in case?
